src/solver/sideWPointSelect.ml

```diff
@@ -0,0 +1,185 @@
+        by the solver.
+    *)
+    val record_destabilized_vs: bool
+    (** This strategy can decide to prevent widening.
+        Note that, if this strategy does not veto, this does not mean that widening
+        will necessarily be performed. Nor does a call to this function imply that
+        the value of the leaf has grown.
+        @param data      The internal state of this strategy
+        @param called    Set of called unknowns
+        @param old_sides Prior side-effects to leaf y
+        @param x         Optional source of the side-effect
+        @param y         Side-effected leaf y
+        @return [true]: widening will not be applied; [false]: widening might be applied
+    *)
+    val veto_widen: data -> unit HM.t -> VS.t -> S.v option -> S.v -> bool
+    (** The value of the leaf has grown. Should it be marked a widening point?
+        Widening points are widened when their value grows, unless vetoed.
+        Even if this function is called, leaf y might already be a widening point
+        from an earlier side-effect.
+        @param data            The internal state of this strategy
+        @param called          Set of called unknowns
+        @param old_sides       Prior side-effects to leaf y
+        @param x               Optional source of the side-effect
+        @param y               Side-effected leaf y
+        @param destabilized_vs Optional destabilization info, described in [record_destabilized_vs]
+        @return [true]: mark as widening point; [false]: do not mark as widening point
+
+    *)
+    val should_mark_wpoint: data -> unit HM.t -> VS.t -> S.v option -> S.v -> bool option -> bool
+  end
+module Always : S =
+  functor (S:EqConstrSys) ->
+  functor (HM:Hashtbl.S with type key = S.v) ->
+  functor (VS:Set.S with type elt = S.v) ->
+  struct
+    type data = unit
+
+    let create_data _ _ = ()
+    let notify_side _ _ _ = ()
+    let record_destabilized_vs = false
+    let veto_widen _ _ _ _ _ = false
+    let should_mark_wpoint _ _ _ _ _ _ = true
+  end
+module Never : S =
+  functor (S:EqConstrSys) ->
+  functor (HM:Hashtbl.S with type key = S.v) ->
+  functor (VS:Set.S with type elt = S.v) ->
+  struct
+    type data = unit
+
+    let create_data _ _ = ()
+    let notify_side _ _ _ = ()
+    let record_destabilized_vs = false
+    let veto_widen _ _ _ _ _ = false
+    let should_mark_wpoint _ _ _ _ _ _ = false
+  end
+module SidesLocal : S =
+  functor (S:EqConstrSys) ->
+  functor (HM:Hashtbl.S with type key = S.v) ->
+  functor (VS:Set.S with type elt = S.v) ->
+  struct
+    type data = unit
+
+    let create_data _ _ = ()
+    let notify_side _ _ _ = ()
+    let record_destabilized_vs = false
+    let veto_widen state called old_sides x y =
+      match x with
+      | None -> false
+      | Some x when VS.mem x old_sides -> false
+      | _ -> true
+    let should_mark_wpoint _ _ _ _ _ _ = true
```

Comment:
Why is this `true` if `sides-local` just had `()` in the old match?

Reply:
In the old implementation, sides-local would not use the `wpoint` set at all for side-effects. Instead, it would only widen if there is an incoming side-effect that increases the global and there had already been a side-effect from the same source. This is very different behavior to the other strategies, which just mark a global as `wpoint` and from then on widen on all increasing contributions.

To put these two behind the same interface, I decided to have the `veto_widen` function. Now, sides-local immediately marks any increasing global as a `wpoint`, but will veto widening whenever the source of the side-effect has not contributed before.
Alternatively, one could replace the `veto_widen` function with something like `widen_anyway`. Then sides-local would not need to mark anything in `wpoint`.